### DISCUSSION TITLE
[Question] Return the query root in a mutation response payload

### DISCUSSION THREAD

**Discussion by rocketraman** [ExpediaGroup/graphql-kotlin#1169] [Q&A] [answered]
I would like to return the query root in a mutation response payload, as described here: https://graphql-rules.com/rules/mutation-payload-query.

Is there a way to do this? I tried returning the value of `DataFetchingEnvironment.getRoot()` on a `Query` type, but that didn't work.

**Comment by smyrick** [comment]
Hmm, interesting. I have not actually seen this pattern before.

I did a quick update of the example and added this function: 

```kotlin
@Component
class SimpleMutation(private val simpleQuery: SimpleQuery) : Mutation {

    fun mutationQueryResponse(input: String): SimpleQuery {
        // Do mutation actions with input
        return simpleQuery
    }
}
```

`SimpleQuery` is marked with `Query` and is a Spring Component. The response type of `mutationQueryResponse` is `SimpleQuery` though and not the global `Query` root object. This is because the schema generator actually handles the creation of the `Query` object for you and there isn't a single object today that you could create and return to get the entire object.

I wonder though if we could do something like allowing to return `com.expediagroup.graphql.server.operations.Query` as a return type and that would return the root object for you. I don't know if graphql-java would support this either.

We usually have a specific Mutation response that you want to return so that you can better control what types of data the clients have access to after they do a mutation, but I can see the benefits of this pattern as well

![Screen Shot 2021-05-29 at 11 21 04 AM](https://user-images.githubusercontent.com/2446877/120080983-f8065700-c06f-11eb-98d3-48e4b9261206.png)

**Reply by smyrick** [reply]
Doing a quick code dive this would require some pretty big changes to the schema generator and server setup.

* Currently we have no single object in `graphql-kotlin-schema-generator` that we can use to represent the `Query` so we may have to move the marker interfaces to the generator.
* If we had the above then we would have to map the usage of the operation back to the internal `GraphQLObjectType` that is the query, cache it in the `SchemaGenerator`, and then return that when we saw the new marker as a return type in Mutations only. We couldn't really do a `Query` in a `Query` as that would put us in a construction loop.

**Reply by smyrick** [reply]
I was actually able to throw something together without a breaking change. We actually don't want to use a marker interface as we need to return a created object to satisfy other parts of the generator. So similar to other libraries, you would just return this empty object and not do anything extra

https://github.com/ExpediaGroup/graphql-kotlin/pull/1170

**Reply by rocketraman** [reply]
Nice!

**Comment by dariuszkuc** [comment]
Given a schema 

```graphql
type Query {
  foo: Foo
  bar: Bar
  baz: Bar
}
```

I am unsure how useful it would be to return whole `Query` type on mutations. 

```graphql
type Mutation {
  updateFoo: UpdatedFoo
}

type UpdatedFoo {
  result: Boolean
  updated: Query
}
```

Guess it really depends on your schema but geneally instead of the whole `Query` object I'd return individual fields that correspond to the target mutation. 

Since `graphql-kotlin` generates schema from source code you could easily return same type for your queries and mutations.

**Reply by smyrick** [reply]
I agree that the pattern itself is not something that I would encourage schema developers to use. While GraphQL does allow clients to run their own queries and free them from API limitation, having a some boundaries for the schema and what is available does make the server and schema maintenance easier.

However, from a library perspective, it appears we can support it and it is supported using an SDL-first approach

**Reply by rocketraman** [reply]
> I am unsure how useful it would be to return whole Query type on mutations.

The source for this is https://graphql-rules.com/rules/mutation-payload-query.

As I understand it, the argument the authors of that site are making for returning the query type is that is is very easy for the client to update its local cache and UI to reflect changes to the graph post-mutation, without an extra round-trip. I see the counter-argument that the schema developer should have already thought through what those changes might be, and exposed those specific elements in the mutation payload, but I guess the root level query there is an escape hatch for the frontend dev in case of a scenario that the backend devs have not considered, or may be in the process of adding/updating.

I thought it was an interesting approach, and I'm not really seeing any concrete reasons in the comments above as to why this would be an anti-pattern or would make schema maintenance any more difficult. One reason might be that it makes it possible for a schema developer to shirk their responsibility to understand what parts of the graph a mutation could change, and return those directly, but I don't see that as very compelling.

**Reply by dariuszkuc** [reply]
While it is possible to do it I don't think it is a good idea. Returning whole `Query` object would only make sense for some very specific and very small GraphQL (micro)services that only have single/few related query fields. Even in those cases I'd argue that it would be more meaningful to return specific `Foo` object vs `Query` containing `Foo`.

I believe the intent with the linked article was that mutations return mutated objects but I would not return complete query object.

**Reply by rocketraman** [reply]
> Returning whole Query object would only make sense for some very specific and very small GraphQL (micro)services that only have single/few related query fields.

Why? Just because the query object is available in the schema doesn't mean the client has to traverse the whole thing. The client can choose to query the specific parts of the graph they are interested in post-mutation.

> I believe the intent with the linked article was that mutations return mutated objects but I would not return complete query object.

No, the linked article specifically argues for returning the top-level `query` in addition to the mutated objects.

**Reply by rocketraman** [reply]
The other approach to querying the back-end post mutation, without an extra network round-trip, would be to batch a mutation and query request, but as I understand it, the GraphQL spec doesn't specify how batched requests are to be dealt with by a server. The server could process them serially in-order, serially out-of-order, or in parallel. Some clients and/or servers may not even support batching, or if they do, they may not allow a mutation and query to be batched together.

This approach, while it feels "dirty" in a sense of allowing the client to query stuff unrelated to the mutation, seems to solve a real technical problem of how to update client-side state without introducing any more round trips, while also somewhat decoupling the mutation business logic from the schema.

**Reply by rocketraman** [reply]
BTW, I created issue https://github.com/graphql-rules/graphql-rules/issues/61 because I'm interested in the author's take on this. I may not be presenting their case accurately.

**Reply by smyrick** [reply]
@dariuszkuc and I had a discussion and while other libraries may support this today, I am not sure if they intentionally did so. Similar to our code, they probably have a recursive parser that just runs the same code for all field definitions and the type defined in the SDL as `Query` is still a valid `GraphQLObjectType` so it works.

But even as called out in the website you posted, it appears magically, and to us that is an anti-pattern of a schema-based API where everything is explicitly defined and the execution code is a 1-1 mapping. We could support this but we think this is actually a strong enough anti-pattern that we do not even want to have the ability, so it is not encouraged.

The spec calls out that the Root Operation Types define the types available to the root of the operation
https://spec.graphql.org/draft/#sec-Root-Operation-Types

Instead we encourage schema developers to think about the data clients need in the mutation response so they don't have to make another round trip and return that data, but not the entire `Query` object. That will ensure the data returned is updated properly by mutation code and there are no other issues with caches and data loaders on other top level fields, as those should be run as a separate operation.


However....if you absolutely wanted to support this case, I think it could be possible if your schema generator hooks maintained some state like my example PR and kept the root query object around, and when you saw a return type of your own custom `class QueryRoot`, you returned your cached `GraphQLObjectType`

**Reply by rocketraman** [reply]
Thank you for the detailed response @smyrick . I'm convinced, but I'd still hoping to hear from the authors of that project about their take. Do you mind if I post your response on the issue above?

**Reply by smyrick** [reply]
Of course, feel free to link to this discussion. I would love to get some other opinions as we don't want to be the sole idea of how a GraphQL schema should be designed, like graphql-rules

**Reply by koenpunt** [reply]
I think the problem of fetching data in a single request could be solved by sending a batch of queries;

Instead of a single query request, which would look something like this:

```json
{
    "query": "mutation { likePost(id: 1) { record { id title likes } query { lastActivePosts { id title likes } } }"
}
```

You would send a request with multiple queries;

```json
[
    {
        "query": "mutation { likePost(id: 1) { record { id title likes } }"
    },
    {
        "query": "query { lastActivePosts { id title likes } }"
    }
]
```

Which I believe should be executed in order, so the second query should retrieve the updated data.

**Reply by rocketraman** [reply]
I discussed the limitations / problems with the query batching approach above: https://github.com/ExpediaGroup/graphql-kotlin/discussions/1169#discussioncomment-812275.

**Reply by rocketraman** [reply]
For posterity, a member of graphql-rules replied here: https://github.com/graphql-rules/graphql-rules/issues/61#issuecomment-855448937.

**Comment by smyrick** [comment, marked as answer]
Update on this, with the addition of `@GraphQLType` you can now just set the return type to a type reference and this works

```kotlin
@GraphQLType("Query")
fun doSomethingAndGetQuery(): Any {
      return Any()
}
```

![Uploading Screen Shot 2022-10-14 at 1.31.52 PM.jpg…]()

**Reply by rocketraman** [reply]
Neat @smyrick , that works very nicely. For the record, this can also be set on a property e.g. following the example given at https://graphql-rules.com/rules/mutation-payload-query:

```
data class LikePostPayload(
  val record: Post,
  @GraphQLType("Query") val query: Any = Any(),
)
```
